[ADP-3306] Make additional API errors machine-readable.

paweljakubas opened the pull request:
This PR makes the following API errors machine-readable:
- `ApiErrorNoSuchPool`
- `ApiErrorMissingWitnessesInTransaction`

In addition, we adjust the integration test suite to express expectations in terms of rich error objects instead of interpolated error message strings.

### Issue

ADP-3306

paweljakubas committed:
Shelley TransactionNew part 1

paweljakubas committed:
Shelley TransactionNew part 2

paweljakubas committed:
refactor MissingWitsInTransaction

paweljakubas committed:
use MinWithdrawalWrong

paweljakubas committed:
refactor NoSuchPool

jonathanknowles committed:
Fix formatting in `IsServerError` instances.

jonathanknowles committed:
Extract out `message` definitions to avoid multi-line parentheses.

jonathanknowles committed:
Fix grammar in `IsServerError` instance for `ErrSubmitTransaction`.

jonathanknowles committed:
Fix formatting in `Cardano.Wallet.Api.Types.Error`.

jonathanknowles committed:
Add strictness annotations to fields of `ApiErrorMissingWitnessesInTransaction`.

jonathanknowles committed:
Add JSON golden test for `ApiErrorMissingWitnessesInTransaction`.

jonathanknowles committed:
Add JSON golden test for `ApiErrorNoSuchPool`.

jonathanknowles committed:
Add API definition for `info` field of `errNoSuchPool`.

jonathanknowles committed:
Add API definition for `info` field of `errMissingWitnessesInTransaction`.

jonathanknowles committed:
Reorder imports into alphabetical order.

jonathanknowles reviewed (commented):
Hi @paweljakubas

This looks great!

Before merging though, it's still missing a couple of things, including Swagger definitions for the new `info` fields, and JSON golden tests.

I've added the missing things in: https://github.com/cardano-foundation/cardano-wallet/pull/4592

jonathanknowles committed:
Regenerate JSON golden file for `ApiError`.

This is necessary to avoid failures of the following type:

```
  src/Test/Aeson/Internal/GoldenSpecs.hs:77:5:
  1) Cardano.Wallet.Api.Types, JSON golden roundtrip, JSON encoding of ApiError, produces the same JSON as is found in /home/jsk/projects/input-output-hk/cardano-wallet-0/lib/unit/test/data/Cardano/Wallet/Api/ApiError.json
       uncaught exception: AesonDecodeError
       AesonDecodeError "Error in $.samples[14]: parsing Cardano.Wallet.Api.Types.Error.ApiErrorInfo(MissingWitnessesInTransaction) failed, key \"info\" not found"

  To rerun use: --match "/Cardano.Wallet.Api.Types/JSON golden roundtrip/JSON encoding of ApiError/produces the same JSON as is found in /home/jsk/projects/input-output-hk/cardano-wallet-0/lib/unit/test/data/Cardano/Wallet/Api/ApiError.json/" --seed 1949888530

Randomized with seed 1949888530
```

The above error appears because we recently added a non-optional `info`
field to the `errMissingWitnessesInTransaction` error.

paweljakubas committed:
final code cleaning

jonathanknowles commented on lib/integration/framework/Test/Integration/Framework/TestData.hs:
```diff
@@ -328,12 +312,6 @@ versionLine = "Running as " <> pack (showFullVersion version gitRevision)
-errMsg403InvalidConstructTx =
-    "It looks like I've created an empty transaction that does not have \
-     \any payments, withdrawals, delegations, metadata nor minting. \
-     \Include at least one of them."
```
🍾

jonathanknowles reviewed (approved):
LGTM! 🎆
